gflownet/policy/multihead_tree.py

```diff
@@ -0,0 +1,502 @@
+from typing import Optional
+
```

Comment:
Note: I have taken a look at the structure of the class and some details but I haven't done a detailed review of all, because I prefer to review the rest of the PR first and delve into this later (after merging, I think).